cmd/metrics.go

```diff
@@ -0,0 +1,129 @@
+/*
+ * Minio Cloud Storage, (C) 2018 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	httpRequests = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "minio_http_requests_total",
+			Help: "Total number of requests served by current Minio server instance",
+		},
+		[]string{"request_type"},
+	)
+	httpRequestsDuration = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "minio_http_requests_duration_seconds",
+			Help:    "Time taken by requests served by current Minio server instance",
+			Buckets: []float64{.001, .003, .005, .1, .5, 1},
+		},
+		[]string{"request_type"},
+	)
+	networkSentBytes = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Name: "minio_network_sent_bytes",
+			Help: "Total number of bytes sent by current Minio server instance",
+		},
+	)
+	networkReceivedBytes = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Name: "minio_network_received_bytes",
+			Help: "Total number of bytes received by current Minio server instance",
+		},
+	)
+	onlineMinioDisks = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Name: "minio_online_disks_total",
+			Help: "Total number of online disks for current Minio server instance",
+		},
+	)
+	offlineMinioDisks = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Name: "minio_offline_disks_total",
+			Help: "Total number of offline disks for current Minio server instance",
+		},
+	)
+	serverUptime = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Name: "minio_server_uptime_seconds",
```

Comment:
There a couple problems here. 

First, this should already be provided by the standard client golang registry as `process_uptime_seconds`.

First, we recommend using a "start time", rather than "up time".  This has several advantages over an uptime ticker.
* You can detect crash loops with something like `changes(process_uptime_seconds[30m])`.
* You can also get the uptime with `time() - process_uptime_seconds`.
* From a storage compression standpoint, a constant value is the easiest to encode.  Prometheus compresses these down extremely well.